### PR TITLE
PackZip

### PR DESCRIPTION
**Pull request by roland-d**
This adds zip support to Robo.li. This is the initial version of PackZip 
and is modeled after PackPhar for consistency.

Please test and review if it works OK for you. I have included a PHPDoc 
that explains how to use the PackZip.

Looking forward to your feedback.

### CODE REVIEW

**Comment by DavertMik** [comment]
Sorry, I just got my hands on it but it looks really cool! Let me test it

File: src/Task/Development/PackZip.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+use Robo\Result;
+use Robo\Task\BaseTask;
+use Symfony\Component\Console\Helper\ProgressBar;
+
+/**
+ * Creates a zip file.
+ *
+ * ``` php
+ * <?php
+ * $this->PackZip(
```
Comment (DavertMik) [inline src/Task/Development/PackZip.php]: there should be a trait created for this task, so you can use it from Robofile like `$this->taskPackZip`.

File: src/Task/Development/PackZip.php [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+ * <full path to folder to zip>,
```
Comment (DavertMik) [inline src/Task/Development/PackZip.php]: why only full path is required? Can we use a path relative to current Robofile?

File: src/Task/Development/PackZip.php [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		$file = new \SplFileInfo($zipname);
+		$this->zipname = $zipname;
+
+		if (file_exists($file->getRealPath()))
+		{
+			@unlink($file->getRealPath());
+		}
+
+		// Instantiate the ZipArchive
+		$this->zip = new \ZipArchive($file->getRealPath(), \ZipArchive::CREATE);
```
Comment (DavertMik) [inline src/Task/Development/PackZip.php]: there should be a check that `zip` extension is enabled

File: src/Task/Development/PackZip.php [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+		// Set all separators to forward slashes for comparison
+		$this->folder = substr(str_replace('\\', '/', $folder), 0, -1);
+
+		// Prepare the zipname
+		$file = new \SplFileInfo($zipname);
+		$this->zipname = $zipname;
+
+		if (file_exists($file->getRealPath()))
+		{
+			@unlink($file->getRealPath());
```
Comment (DavertMik) [inline src/Task/Development/PackZip.php]: any *real* actions that may affect changes should be done in `run` method

File: src/Task/Development/PackZip.php [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		// Inform the user which archive we are creating
+		$this->printTaskInfo("creating <info>{$this->zipname}</info>");
+
+		// Instantiate the zip archive
+		$this->zip->open($this->zipname, \ZipArchive::CREATE);
+
+		// Count the files to pack
+		$files = 0;
+
+		foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->folder), \RecursiveIteratorIterator::SELF_FIRST) as $subfolder)
```
Comment (DavertMik) [inline src/Task/Development/PackZip.php]: Actually you can use Symfony Finder here. It is included

File: src/Task/Development/PackZip.php [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	 *
+	 * @since   1.0
+	 */
+	public function run()
+	{
+		// Inform the user which archive we are creating
+		$this->printTaskInfo("creating <info>{$this->zipname}</info>");
+
+		// Instantiate the zip archive
+		$this->zip->open($this->zipname, \ZipArchive::CREATE);
```
Comment (DavertMik) [inline src/Task/Development/PackZip.php]: should check for errors and return `Result::error` if something wrong happened

File: src/Task/Development/PackZip.php [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+					$addpath = substr($addpath, 1);
+				}
+
+				if (strlen($addpath) > 0 || empty($addpath))
+				{
+					$addpath .= '/';
+				}
+
+				$options = array('add_path' => $addpath, 'remove_all_path' => true);
+				$this->zip->addGlob($usefolder . '/*.*', GLOB_BRACE, $options);
```
Comment (DavertMik) [inline src/Task/Development/PackZip.php]: should check for errors and return Result::error if something wrong happened

**Comment by burzum** [comment]
@roland-d any news from your side on this, you've got your feedback. :)

**Comment by roland-d** [comment]
@burzum I haven't forgotten about this for sure, just haven't had time to get to it. With the busiest time of the year behind is, I will be looking into this soon. Thanks.

**Comment by greg-1-anderson** [comment]
I recommend closing this in favor of the implementation in #244.

**Comment by roland-d** [comment]
Thanks for that @greg-1-anderson Closing this one.

**Pull request closed without merging**